File: modules/core/middleware/sentry.go
package middleware

import (
	"log/slog"
	"os"

	"github.com/getsentry/sentry-go"
	sentrygin "github.com/getsentry/sentry-go/gin"
	"github.com/gin-gonic/gin"
)

// InitSentry initialises the Sentry SDK from the SENTRY_DSN environment variable.
// If SENTRY_DSN is empty, Sentry is disabled gracefully — no error is returned.
// PII scrubbing: Authorization and Cookie headers are removed from all events.
func InitSentry(logger *slog.Logger) error {
	dsn := os.Getenv("SENTRY_DSN")
	if dsn == "" {
		logger.Info("SENTRY_DSN not set, Sentry error tracking disabled")
		return nil
	}

	env := os.Getenv("GIN_MODE")
	if env == "" {
		env = "development"
	}

	err := sentry.Init(sentry.ClientOptions{
		Dsn:              dsn,
		Environment:      env,
		TracesSampleRate: 0.1,
		BeforeSend: func(event *sentry.Event, hint *sentry.EventHint) *sentry.Event {
			if event.Request != nil {
				event.Request.Cookies = ""
				delete(event.Request.Headers, "Authorization")
				delete(event.Request.Headers, "Cookie")
			}
			return event
		},
	})
	if err != nil {
		return err
	}

	logger.Info("Sentry error tracking initialised", "environment", env)
	return nil
}

// SentryMiddleware returns the sentrygin middleware configured to re-panic after
// capturing so that gin.Recovery() can still handle the panic response.
func SentryMiddleware() gin.HandlerFunc {
	return sentrygin.New(sentrygin.Options{Repanic: true})
}

File: services/frontend/src/lib/analytics/sentry.ts
import { browser } from '$app/environment';

let sentryLoaded = false;

export async function initSentry(): Promise<void> {
	if (!browser) return;

	const dsn = import.meta.env.VITE_SENTRY_DSN;
	if (!dsn) return;

	try {
		const Sentry = await import('@sentry/sveltekit');

		Sentry.init({
			dsn,
			environment: import.meta.env.MODE,
			release: `matchup@${import.meta.env.VITE_APP_VERSION ?? 'dev'}`,
			tracesSampleRate: 0.1,
			beforeSend(event) {
				if (event.request?.headers) {
					delete event.request.headers['Authorization'];
					delete event.request.headers['Cookie'];
				}
				if (event.request) {
					delete event.request.cookies;
				}
				return event;
			}
		});

		sentryLoaded = true;
	} catch (e) {
		console.warn('[Sentry] Failed to initialise:', e);
	}
}

export async function setSentryUser(userId: string): Promise<void> {
	if (!browser || !sentryLoaded) return;
	const Sentry = await import('@sentry/sveltekit');
	Sentry.setUser({ id: userId });
}

export async function clearSentryUser(): Promise<void> {
	if (!browser || !sentryLoaded) return;
	const Sentry = await import('@sentry/sveltekit');
	Sentry.setUser(null);
}
